### PR TITLE
Add option to retry on hosted form validation request returning HTTP 5xx

### PR DESCRIPTION
**Pull request by horgh**
This adds two new attributes to `SecureToken`.One to set how many times we will retry if we see a HTTP 5xx response. The second is a callback we call prior to retrying. This callback is useful for logging.

### CODE REVIEW

File: Changes [hunk 1]
```diff
@@ -2,6 +2,11 @@ Revision history for WebService-PayPal-PaymentsAdvanced
 
 {{$NEXT}}
 
+0.000022 2017-05-08
```
Comment (oschwald) [inline Changes]: This line is unnecessary. It is automatically generated by dzil on release. We should probably do a release when this PR is merged.

File: lib/WebService/PayPal/PaymentsAdvanced.pm [hunk 2]
```diff
@@ -601,6 +604,11 @@ to C<true>.
 Create a secure token which you can use to create a hosted form uri.  Returns a
 L<WebService::PayPal::PaymentsAdvanced::Response::SecureToken> object.
 
+The first parameter holds the key/value parameters for the request. The second
+parameter is optional, and holds parameters to the underlying
```
Comment (oschwald) [inline lib/WebService/PayPal/PaymentsAdvanced.pm]: No comma after "optional".

File: lib/WebService/PayPal/PaymentsAdvanced/Response/SecureToken.pm [hunk 3]
```diff
@@ -82,6 +100,56 @@ sub _build_hosted_form_uri {
...
+
+    my $res;
+
+    # +1 so we always try at least once.
+    ## no critic (ControlStructures::ProhibitCStyleForLoops)
+    for (
+        my $attempts_left = $self->retry_attempts + 1;
+        $attempts_left > 0;
+        $attempts_left--
+        ) {
```
Comment (oschwald) [inline lib/WebService/PayPal/PaymentsAdvanced/Response/SecureToken.pm]: Any reason to not just do `for my $attempt (1..$self->retry_attempts+1)` here and adjust the `last` check appropriately? The C-style for loop is a bit harder to parse and more prone to off-by-one errors.

File: lib/WebService/PayPal/PaymentsAdvanced/Response/SecureToken.pm [hunk 4]
```diff
@@ -109,6 +177,17 @@ The URL for the PayflowLink web service.  Can be a mocked URL.
...
+=head2 retry_attempts
+
+The number of HTTP retries to attempt if we encounter an error response. We
+retry only when encountering HTTP 5xx responses.
+
+=head2 retry_callback
+
+The callback to call if we encounter an HTTP 5xx response. We call this prior to
+retrying the request. The callback will have a single parameter, an
+HTTP::Response object.
```
Comment (oschwald) [inline lib/WebService/PayPal/PaymentsAdvanced/Response/SecureToken.pm]: From this, it isn't clear what I am supposed to do with the callback or what its purpose is. A sentence about the motivation or with a use case would go a long way in helping the reader understand.

File: t/PaymentsAdvanced/Response/SecureToken.t [hunk 5]
```diff
@@ -1,3 +1,5 @@
+#!/usr/bin/env perl;
```
Comment (oschwald) [inline t/PaymentsAdvanced/Response/SecureToken.t]: This should be needed. The `perlcriticrc` file needed to be updated, but now that you rebased against my changes, it should work fine without.

**Review by oschwald — CHANGES_REQUESTED** [review]
Looks good! I had a couple of minor comments.

File: lib/WebService/PayPal/PaymentsAdvanced/Response/SecureToken.pm (hunk 3 above)
Comment (horgh) [reply lib/WebService/PayPal/PaymentsAdvanced/Response/SecureToken.pm]: The rationale for the C loop was for the `last` check to be clearly about whether this is the last attempt we have left. However I agree that the count down required more thought. In fact I had an off by one error on my first pass writing it! I've updated it now.

**Pull request merged**